### PR TITLE
Sample verifier, LOG Sprint 10

### PR DESCRIPTION
**Pull request by stefangenchev**
- Added the consistency and inclusion proof verification methods to the main classes.
- Added the hash controller API interface.
- Added a test that verifies the returned consistency and inclusion proofs, the MTH time stamp and the hash chain between two entries

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -22,3 +22,66 @@ build
 .classpath
 .project
 .settings/
+
+# Covers JetBrains IDEs: IntelliJ, RubyMine, PhpStorm, AppCode, PyCharm, CLion, Android Studio and WebStorm
+# Reference: https://intellij-support.jetbrains.com/hc/en-us/articles/206544839
+
+# User-specific stuff
```
Comment (Glamdring) [inline .gitignore]: should be just .idea/ and everything underneath will be ignored.

File: logsentinel-java-client.iml [hunk 2]
```diff
@@ -0,0 +1,41 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (Glamdring) [inline logsentinel-java-client.iml]: this file should be added to .gitignore

File: src/main/java/com/logsentinel/LogSentinelClient.java [hunk 3]
```diff
@@ -39,4 +41,6 @@ public ManageApplicationControllerApi getApplicationActions() {
     public AuditLogSearchControllerApi getSearchActions() {
         return searchActions;
     }
+
+    public ApiVerificationControllerApi getVerificationActions() { return verificationActions; }
```
Comment (Glamdring) [inline src/main/java/com/logsentinel/LogSentinelClient.java]: spread on multiple lines

File: src/main/java/com/logsentinel/merkletree/utils/MathUtils.java [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+public class MathUtils {
+    private static final double logSubtractFromNumber = 0.01;
+
+    /**
+     * Calculates the parameter k - the largest power of 2 smaller than n (i.e., k &lt; n &lt;= 2k)
+     *
+     * @param n the parameter n that is used to calculate the parameter k
+     * @return the calculated parameter k
+     */
+    public static int calculateK(int n) {
```
Comment (piro314) [inline src/main/java/com/logsentinel/merkletree/utils/MathUtils.java:19]: Too many Math.log() is not very effective. Consider using something like:
Math.pow(2, (Integer.toBinaryString(n).length() -1 ))

File: src/main/java/com/logsentinel/client/ApiVerificationControllerApi.java [hunk 5]
```diff
@@ -0,0 +1,829 @@
...
+    }
+
+    /**
+     * Get all resolved verification failures
+     * 
+     * @param applicationId Application ID, identifying a target application (obtained from the API credentials page) (required)
+     * @return CollectionOfResolvedVerificationFailureResponse
+     * @throws ApiException If fail to call the API, e.g. server error or cannot deserialize the response body
+     */
+    public CollectionOfResolvedVerificationFailureResponse getResolvedVerificationFailuresUsingGET(String applicationId) throws ApiException {
```
Comment (Glamdring) [inline src/main/java/com/logsentinel/client/ApiVerificationControllerApi.java]: this return type is not okay. I don't know why it was generated like that, but it should be a List

File: src/main/java/com/logsentinel/client/AuditLogControllerApi.java [hunk 6]
```diff
@@ -62,6 +62,143 @@ public void setApiClient(ApiClient apiClient) {
...
+    /**
+     * Build call for getEntryByIdUsingGET
+     * @param applicationId Application ID, identifying a target application (obtained from the API credentials page) (required)
+     * @param entryId ID of the audit log entry to verify (required)
+     * @param progressListener Progress listener
+     * @param progressRequestListener Progress request listener
+     * @return Call to execute
+     * @throws ApiException If fail to serialize the request body object
+     */
+    public com.squareup.okhttp.Call getEntryByIdUsingGETCall(String applicationId, String entryId, final ProgressResponseBody.ProgressListener progressListener, final ProgressRequestBody.ProgressRequestListener progressRequestListener) throws ApiException {
```
Comment (Glamdring) [inline src/main/java/com/logsentinel/client/AuditLogControllerApi.java]: get rid of "UsingGETCall" (UsingXXX in general should be remvoed)

File: src/main/java/com/logsentinel/client/model/CollectionOfResolvedVerificationFailureResponse.java [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+
+package com.logsentinel.client.model;
+
+import java.util.Objects;
+
+/**
+ * CollectionOfResolvedVerificationFailureResponse
+ */
+@javax.annotation.Generated(value = "io.swagger.codegen.languages.JavaClientCodegen", date = "2018-07-02T11:34:18.314+03:00")
+public class CollectionOfResolvedVerificationFailureResponse {
```
Comment (Glamdring) [inline src/main/java/com/logsentinel/client/model/CollectionOfResolvedVerificationFailureResponse.java]: delete this class altogether

File: src/main/java/com/logsentinel/client/model/LogInfo.java [hunk 8]
```diff
@@ -0,0 +1,31 @@
+package com.logsentinel.client.model;
+
+public class LogInfo {
```
Comment (Glamdring) [inline src/main/java/com/logsentinel/client/model/LogInfo.java]: isn't this the same as MerkleTreeInfo? If yes, leave only one

File: src/main/java/com/logsentinel/client/model/MerkleTreeInfo.java [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+ */
+@javax.annotation.Generated(value = "io.swagger.codegen.languages.JavaClientCodegen", date = "2018-07-02T11:34:18.314+03:00")
+public class MerkleTreeInfo {
+  @SerializedName("hashAlgorithmOid")
+  private String hashAlgorithmOid = null;
+
+  @SerializedName("leafHashAlgorithmOid")
+  private String leafHashAlgorithmOid = null;
+
+  @SerializedName("publicKey")
```
Comment (Glamdring) [inline src/main/java/com/logsentinel/client/model/MerkleTreeInfo.java]: I think the name was changed

File: src/main/java/com/logsentinel/merkletree/generation/TreeHeadGeneration.java [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+
+/**
+ * Merkle Tree Hash / Head (MTH) generation
+ * MODIFIED FOR CLIENT
+ *
+ * @author Stefan Genchev
+ * @author https://stefan.genchev.io/
+ * @version 1.0
+ */
+@SuppressWarnings("SpellCheckingInspection")
```
Comment (Glamdring) [inline src/main/java/com/logsentinel/merkletree/generation/TreeHeadGeneration.java]: I think that's not needed

File: src/main/java/com/logsentinel/merkletree/verification/InclusionProofVerification.java [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+import java.util.List;
+
+/**
+ * Merkle inclusion proof verification
+ *
+ * @author Stefan Genchev
+ * @author https://stefan.genchev.io/
+ * @version 1.0
+ */
+public class InclusionProofVerification {
```
Comment (Glamdring) [inline src/main/java/com/logsentinel/merkletree/verification/InclusionProofVerification.java]: rename the two classes from XXVerification to XXVerifier

Comment (Glamdring) [reply src/main/java/com/logsentinel/merkletree/verification/InclusionProofVerification.java]: (use refactoring, of course :) )

File: src/main/java/com/logsentinel/util/StringUtil.java [hunk 12]
```diff
@@ -52,4 +52,23 @@ public static String join(String[] array, String separator) {
...
         return out.toString();
     }
+
+    /**
+     * Add the missing padding of a Base-64 string.
+     *
+     * @param base64 Base64-encoded string
+     * @return padded Base-64 string
+     */
+    public static String base64StringAddPadding(String base64) {
```
Comment (Glamdring) [inline src/main/java/com/logsentinel/util/StringUtil.java:62]: why do we need this? The server-returned base64 should be okay. Or it isn't?

Comment (stefangenchev) [reply src/main/java/com/logsentinel/util/StringUtil.java:62]: The server returns Base-64 encoded data without the needed padding.

File: src/main/java/com/logsentinel/merkletree/utils/MathUtils.java (hunk 4 above)
Comment (stefangenchev) [reply src/main/java/com/logsentinel/merkletree/utils/MathUtils.java:19]: Optimization of the code sample is not necessary.

File: src/main/java/com/logsentinel/client/model/LogInfo.java (hunk 8 above)
Comment (stefangenchev) [reply src/main/java/com/logsentinel/client/model/LogInfo.java]: The wrong swagger API definition for the retrieval of basic log information caused confusion (wrong return type) and I did not found the appropriate class. But there it is. Going to fix this.

File: src/main/java/com/logsentinel/client/model/MerkleTreeInfo.java (hunk 9 above)
Comment (stefangenchev) [reply src/main/java/com/logsentinel/client/model/MerkleTreeInfo.java]: Yes, it was changed to timestampPublicKey.

**Review by Glamdring — APPROVED** [review]

**Pull request merged**